README.md
# NAME

Mojolicious::Plugin::Webpack - Mojolicious ♥ Webpack

# SYNOPSIS

## Define entrypoint

Create a file "./assets/index.js" relative to you application directory with
the following content:

    console.log('Cool beans!');

## Application

Your [Mojolicious](https://metacpan.org/pod/Mojolicious) application need to load the
[Mojolicious::Plugin::Webpack](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3AWebpack) plugin and tell it what kind of assets it
should be able to process:

    $app->plugin(Webpack => {process => [qw(js css)]});

See ["register"](#register) for more configuration options.

## Template

To include the generated assets in your template, you can use the ["asset"](#asset)
helper:

    %= asset "cool_beans.css"
    %= asset "cool_beans.js"

## Run the application

You can start the application using `daemon`, `hypnotoad` or any Mojolicious
server you want, but if you want rapid development you should use
`mojo webpack`, which is an alternative to `morbo`:

    $ mojo webpack -h
    $ mojo webpack ./script/myapp.pl

However if you want to use another daemon and force `webpack` to run, you need
to set the `MOJO_WEBPACK_BUILD` environment variable to "1". Example:

    MOJO_WEBPACK_BUILD=1 ./script/myapp.pl daemon

## Testing

If you want to make sure you have built all the assets, you can make a test
file like "build-assets.t":

    use Test::More;
    use Test::Mojo;

    # Run with TEST_BUILD_ASSETS=1 prove -vl t/build-assets.t
    plan skip_all => "TEST_BUILD_ASSETS=1" unless $ENV{TEST_BUILD_ASSETS};

    # Load the app and make a test object
    $ENV{MOJO_MODE}          = 'production';
    $ENV{MOJO_WEBPACK_BUILD} = 1;
    use FindBin;
    require "$FindBin::Bin/../script/myapp.pl";
    my $t = Test::Mojo->new;

    # Find all the tags and make sure they can be loaded
    $t->get_ok("/")->status_is(200);
    $t->element_count_is('script[src], link[href][rel=stylesheet]', 2);
    $t->tx->res->dom->find("script[src], link[href][rel=stylesheet]")->each(sub {
      $t->get_ok($_->{href} || $_->{src})->status_is(200);
    });

    done_testing;

# DESCRIPTION

[Mojolicious::Plugin::Webpack](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3AWebpack) is a [Mojolicious](https://metacpan.org/pod/Mojolicious) plugin to make it easier to
work with [https://webpack.js.org/](https://webpack.js.org/) or [https://rollupjs.org/](https://rollupjs.org/). This plugin
will...

1. Generate a minimal `package.json` and a Webpack or Rollup config file. Doing
this manually is possible, but it can be quite time consuming to figure out all
the bits and pieces if you are not already familiar with Webpack.
2. Load the entrypoint "./assets/index.js", which is the starting point of your
client side application. The entry file can load JavaScript, CSS, SASS, ... as
long as the appropriate processing plugin is loaded.
3. It can be difficult to know exactly which plugins to use with Webpack. Because
of this [Mojolicious::Plugin::Webpack](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3AWebpack) has some predefined rules for which
Nodejs dependencies to fetch and  install. None of the nodejs modules are
required in production though, so it will only be installed while developing.
4. While developing, the webpack executable will be started automatically next to
[Mojo::Server::Morbo](https://metacpan.org/pod/Mojo%3A%3AServer%3A%3AMorbo). Webpack will be started with the appropriate switches
to watch your source files and re-compile on change.

## Rollup

[rollup.js](https://rollupjs.org/) is an alternative to Webpack. Both
accomplish more or less the same thing, but in different ways.

To be able to use rollup, you have to load this plugin with a different engine:

    $app->plugin(Webpack => {engine => 'Mojo::Alien::rollup', process => [qw(js css)]});

## Notice

[Mojolicious::Plugin::Webpack](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3AWebpack) is currently EXPERIMENTAL.

# HELPERS

## asset

    # Call a method or access an attribute in this class
    my $path = $app->asset->engine->out_dir;

    # Call a method, but from inside a mojo template
    %= asset->url_for($c, "cool_beans.css")

    # Generate a HTML tag
    my $bytestream = $c->asset("cool_beans.js", @args);

    # Generate a HTML tag, but from inside a mojo template
    %= asset "cool_beans.css", media => "print"
    %= asset(url_for => "cool_beans.css")

The most basic usage of this helper is to create a HTML tag using
["javascript" in Mojolicious::Plugin::TagHelpers](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3ATagHelpers#javascript) or
["stylesheet" in Mojolicious::Plugin::TagHelpers](https://metacpan.org/pod/Mojolicious%3A%3APlugin%3A%3ATagHelpers#stylesheet) if a valid asset name is passed in.

On the other hand, the helper will return the plugin instance if no arguments
are passed in, allowing you to call any of the ["METHODS"](#methods) or access the
["ATTRIBUTES"](#attributes).

# HOOKS

## before\_webpack\_start

    $app->before_webpack_start(sub { my $webpack = shift; ... });

Emitted right before the plugin starts building or loading in the generated
assets. Useful if you want to change any of the ["engine"](#engine) attributes from the
defaults.

# ATTRIBUTES

## engine

    $engine = $webpack->engine;

Returns a [Mojo::Alien::webpack](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Awebpack) or [Mojo::Alien::rollup](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Arollup) object.

# METHODS

## asset\_map

    $hash_ref = $webpack->asset_map;

Reads all the generated files in ["asset\_path"](#asset_path) and returns a hash-ref like
this:

    {
      "relative/output.js" => {               # Key is a friendly name, withouc checksum
        ext      => 'css',                    # File extension
        helper   => 'javascript',             # Mojolicious helper used to render the asset
        rel_name => "relative/output.xyz.js", # Relative filename with checksum
      },
      ...
    }

Note that changing this hash might change how ["asset"](#asset) and ["url\_for"](#url_for) behaves!

## register

    $webpack->register($app, \%config);
    $app->plugin("Webpack", \%config);

Used to register this plugin into your [Mojolicious](https://metacpan.org/pod/Mojolicious) app.

The `%config` passed when loading this plugin can have any of these
attributes:

- asset\_path

    Can be used to specify an alternative static directory to output the built
    assets to.

    Default: "/asset".

- cache\_control

    Used to set the response "Cache-Control" header for built assets.

    Default: "no-cache" while developing and "max-age=86400" in production.

- engine

    Must be a valid engine class name. Examples:

        $app->plugin("Webpack", {engine => 'Mojo::Alien::rollup'});
        $app->plugin("Webpack", {engine => 'Mojo::Alien::webpack'});

    Default: [Mojo::Alien::webpack](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Awebpack).

- helper

    Name of the helper that will be added to your application.

    Default: "asset".

- process

    Used to specify ["include" in Mojo::Alien::webpack](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Awebpack#include) or  ["include" in Mojo::Alien::rollup](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Arollup#include).

    Default: `['js']`.

## url\_for

    $url = $webpack->url_for($c, $asset_name);

Returns a [Mojo::URL](https://metacpan.org/pod/Mojo%3A%3AURL) for a given asset.

# AUTHOR

Jan Henning Thorsen

# COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

# SEE ALSO

[Mojo::Alien::rollup](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Arollup), [Mojo::Alien::webpack](https://metacpan.org/pod/Mojo%3A%3AAlien%3A%3Awebpack).


Changes
Revision history for perl distribution Mojolicious-Plugin-Webpack

1.02 Not Released
 - Fix "pnpm init -y" is not a valid command
 - Fix installing both devDependencies and dependencies #14 #15
   Contributor: Daniel Mantovani

1.01 2021-05-25T07:36:24+0900
 - Removed need for assets/asset_map.mode.json
 - Add support for mojo webpack --build

1.00 2021-05-24T15:17:08+0900
 - Add Mojo::Alien::npm
 - Add Mojo::Alien::rollup
 - Add Mojo::Alien::webpack
 - Made Mojo::Alien::npm compatible with https://pnpm.io/
 - Improved dependency tracking in ./node_modules
 - Rewrote Plugin::Webpack using new Mojo::Alien:: modules
 - Updated examples to match "alien" modules
 - Fix being able to include images and other processed assets #7
 - Fix vulnerabilities reported by GitHub #12
 - Fix CI Ubuntu workflow

0.14 2020-12-21T09:42:41+0900
 - Fix route warnings issued by latest Mojolicious
 - Will install new node modules present in "dependencies"
 - Migrated to webpack 5
   Contributor: Daniel Mantovani

0.13 2020-09-07T10:10:24+0900
 - Fix "sass-loader" options #4
 - Removed support for "shim"
 - Improved documentation

0.12 2019-10-07T07:02:11+0900
 - Renamed every ".dev." file to ".development." #2

0.11 2019-10-05T22:12:07+0900
 - Will skip entries in generated HTML file that does not match current mode.

0.10 2019-10-02T15:35:43+0900
 - Add support for custom "npm" binary
 - Need to look in "devDependencies" as well to skip install of node module
 - Will only install modules once

0.09 2019-05-20T22:37:42+0200
 - Add support for rollup.js
 - Changed the EXPERIMENTAL statement

0.08 2019-04-04T22:34:48+0700
 - Fix annoying default for Cache-Control when developing

0.07 2019-04-04T22:19:40+0700
 - Add Cache-Control to response headers
 - Fix passing on arguments using the "webpack" command
 - Change webpack-md5-hash to the superior webpack-plugin-hash-output

0.06 2019-01-13T10:52:34+0900
 - Converted crushinator script to mojo webpack subcommand
   You need to manually remove `which crushinator` if already installed.

0.05 2018-12-27T13:21:34+0900
 - Use terser-webpack-plugin instead of uglifyjs-webpack-plugin

0.04 2018-12-26T13:53:46+0900
 - Fix typo in webpack.custom.js
 - Add M::P::Webpack::Builder to make M::P::Webpack a production shim
 - M::P::Webpack::Builder can install a shim into your application

0.03 2018-12-20T12:50:22+0900
 - Add url_for() method
 - Will migrate assetpack.def from AssetPack into webpack custom config
 - Change webpack to only run when requested
 - Documented how to test your assets

0.02 2018-11-16T19:53:02+0900
 - Fix markup.t

0.01 2018-11-07T15:44:27+0900
 - Will automatically setup webpack config files
 - Can process CSS, JavaScript, Sass and Vue
 - Can extend the webpack config file
 - Add automatic installation of node modules
 - Add crushinator for development


lib/Mojo/Alien/webpack.pm
package Mojo::Alien::webpack;
use Mojo::Base -base;

use Carp qw(croak);
use File::chdir;
use Mojo::Alien::npm;
use Mojo::File qw(path tempfile);
use Mojo::Loader;
use POSIX ':sys_wait_h';
use Time::HiRes qw(sleep);

use constant DEBUG => ($ENV{MOJO_ROLLUP_DEBUG} || $ENV{MOJO_WEBPACK_DEBUG}) && 1;

# TODO
our $VERSION = $Mojolicious::Plugin::Webpack::VERSION || '0.01';

has assets_dir => sub { shift->config->dirname->child('assets') };

has binary => sub {
  my $self = shift;
  return $ENV{MOJO_WEBPACK_BINARY} if $ENV{MOJO_WEBPACK_BINARY};
  my $bin = $self->config->to_abs->dirname->child(qw(node_modules .bin webpack));
  $self->_d('%s %s', -e $bin ? 'Found' : 'Not installed', $bin) if DEBUG;
  return -e $bin ? $bin->to_string : 'webpack';
};

has config => sub { path->to_abs->child('webpack.config.js') };

has dependencies => sub {
  return {
    core   => [qw(webpack webpack-cli)],
    css    => [qw(css-loader mini-css-extract-plugin css-minimizer-webpack-plugin)],
    eslint => [qw(eslint eslint-webpack-plugin)],
    js     => [qw(@babel/core @babel/preset-env @babel/plugin-transform-runtime babel-loader terser-webpack-plugin)],
    sass   => [qw(css-loader mini-css-extract-plugin css-minimizer-webpack-plugin sass sass-loader)],
    vue    => [qw(vue vue-loader vue-template-compiler)],
  };
};

has include => sub { +[] };
has mode    => sub { $ENV{NODE_ENV} || 'development' };

has npm => sub {
  my $self = shift;
  Mojo::Alien::npm->new(config => $self->config->dirname->child('package.json'), mode => $self->mode);
};

has out_dir => sub { shift->config->dirname->child('dist') };

sub asset_map {
  my $self = shift;

  my %assets;
  for my $path ($self->out_dir->list_tree->each) {
    my $rel_path = File::Spec->abs2rel($path, $self->out_dir);
    my $name     = $rel_path;
    $name =~ s!(.*)\W(\w+)\.(\w+)$!$1.$3!;    # (prefix, checksum, ext)
    my $mode = ($2 // '') eq 'development' ? 'development' : 'production';
    $assets{$rel_path} = {ext => lc $3, name => $name, mode => $mode, mtime => $path->stat->mtime, path => $path};
  }

  return \%assets;
}

sub build {
  my $self = shift;
  croak "Can't call build() after watch()" if $self->pid;

  ($!, $?) = (0, 0);
  $self->_run($self->_cmd_build);
  croak "$self->{basename} $! (exit=$?)"   if $!;
  croak "$self->{basename} failed exit=$?" if !$! and $?;

  return $self;
}

sub exec {
  my $self = shift;
  my @cmd  = ($self->_cmd_build, '--watch');
  my $home = $self->config->dirname->to_string;

  chdir $home or die "Can't chdir to $home: $!";
  $ENV{NODE_ENV}           = $self->mode;
  $ENV{WEBPACK_ASSETS_DIR} = $self->assets_dir->to_string;
  $ENV{WEBPACK_OUT_DIR}    = $self->out_dir->to_string;
  $self->_d('(%s) cd %s && %s', $$, $home, join ' ', @_) if DEBUG;
  { CORE::exec(@cmd) }
  die "Can't exec @cmd: $!";
}

sub init {
  my $self = shift;

  $self->npm->init;
  $self->_render_file($self->_config_template_name, $self->config);

  my $dependencies = $self->npm->dependencies;
  my @includes     = @{$self->include};
  push @includes, 'core' unless grep { $_ eq 'core' } @includes;
  my ($conf_d, @includes_names, %seen) = ($self->_config_include_dir);
  for my $include (@includes) {
    for my $package (@{$self->dependencies->{$include} || []}) {
      next if $seen{$package}++;
      $self->npm->install($package) unless $dependencies->{$package}{version};
    }

    my $exists = $self->_resources->{"include/$include.js"} ? 'exists' : 'does not exist';
    my $file   = $conf_d->child("$include.js");
    $self->_d('Template %s.js %s', $include, $exists) if DEBUG;
    $self->_render_file("include/$include.js", $file) if $exists eq 'exists';
    push @includes_names, $include if -e $file;
  }

  my $include_src = ("module.exports = function(config, opts) {\n");
  $include_src .= "  require('./$_')(config, opts);\n" for @includes_names;
  $include_src .= "};\n";
  $self->_render_file('include.js', $self->_config_include_dir->child('include.js'), $include_src);
  return $self;
}

sub pid {
  my $self = shift;
  return 0 unless $self->{pid};
  my $r = waitpid $self->{pid}, WNOHANG;    # -1 == no such process, >0 if terminated
  return $r == -1 && delete $self->{pid} ? 0 : $r ? 0 : $self->{pid};
}

sub stop {
  my ($self, $tries) = @_;

  $tries ||= 100;
  while (--$tries) {
    return $self unless my $pid = $self->pid;
    local $!;
    kill 15, $pid;
    waitpid $pid, 0;
    sleep $ENV{MOJO_WEBPACK_STOP_INTERVAL} || 0.1;
  }

  $self->{basename} ||= path($self->binary)->basename;
  croak "Couldn't stop $self->{basename} with pid @{[$self->pid]}";
}

sub watch {
  my $self = shift;
  return $self if $self->pid;

  my $home = $self->config->dirname->to_string;
  croak "Can't chdir $home: No such file or directory" unless -d $home;

  my @cmd = ($self->_cmd_build, '--watch');
  croak "Can't fork: $!" unless defined(my $pid = fork);
  return $self if $self->{pid} = $pid;    # Parent
  return $self->exec;                     # Child
}

sub _cmd_build {
  my $self = shift;
  $self->init;

  my @cmd = ($self->binary);
  croak "Can't run $cmd[0]" unless -x $cmd[0];

  $self->{basename} ||= path($cmd[0])->basename;
  push @cmd, '--config' => $self->config->to_string;
  push @cmd, qw(--progress --profile --verbose) if $ENV{MOJO_WEBPACK_VERBOSE};
  return @cmd;
}

sub _config_include_dir   { shift->assets_dir->child('webpack.config.d') }
sub _config_template_name {'webpack.config.js'}
sub _d                    { my ($class, $format) = (shift, shift); warn sprintf "[Webpack] $format\n", @_ }

sub _render_file {
  my ($self, $name, $file, $content) = @_;

  if (-e $file) {
    my $version = $file->slurp =~ m!// Autogenerated.*(\d+\.\d+)! ? $1 : -1;
    $self->_d('File %s has version %s', $file, $version) if DEBUG;
    return $self                                         if $version == -1;
    return $self if $version == $VERSION and !$content and !$ENV{MOJO_WEBPACK_REGENERATE};
  }

  $self->_d('Render %s to %s', $name, $file) if DEBUG;
  $file->dirname->make_path unless -d $file->dirname;
  $content //= $self->_resources->{$name};
  $file->spurt(sprintf "// Autogenerated by %s %s\n%s", ref($self), $VERSION, $content);
  return $self;
}

sub _resources {
  state $resources = Mojo::Loader::data_section(ref($_[0]) || $_[0]);
}

sub _run {
  my ($self, @cmd) = @_;
  local $CWD                     = $self->config->dirname->to_string;
  local $ENV{NODE_ENV}           = $self->mode;
  local $ENV{WEBPACK_ASSETS_DIR} = $self->assets_dir->to_string;
  local $ENV{WEBPACK_OUT_DIR}    = $self->out_dir->to_string;
  $self->_d('cd %s && %s', $CWD, join ' ', @cmd) if DEBUG;
  open my $WEBPACK, '-|', @cmd or die "Can't run @cmd: $!";
  return $WEBPACK if defined wantarray;
  DEBUG && print while <$WEBPACK>;
}

sub DESTROY { shift->stop }

1;

=encoding utf8

=head1 NAME

Mojo::Alien::webpack - Runs the external nodejs program webpack

=head1 SYNOPSIS

  use Mojo::Alien::webpack;
  my $webpack = Mojo::Alien::webpack->new;

  # Run once
  $webpack->build;

  # Build when webpack see files change
  $webpack->watch;

=head1 DESCRIPTION

L<Mojo::Alien::webpack> is a class for runnig the external nodejs program
L<webpack|https://webpack.js.org/>.

=head1 ATTRIBUTES

=head2 assets_dir

  $path = $webpack->assets_dir;
  $webpack = $webpack->assets_dir($webpack->config->dirname->child('assets'))

Location to source assetsa and partial webpack.config.js files.

=head2 binary

  $array_ref = $webpack->binary;
  $webpack = $webpack->binary('webpack');

The path to the webpack executable. Defaults to C<MOJO_WEBPACK_BINARY>
environment variable, or "webpack" inside "./node_modules". Fallback to just
"webpack".

=head2 config

  $path = $webpack->config;
  $webpack = $webpack->config(path->to_abs->child('webpack.config.js'));

Holds an I</absolute> path to
L<webpack.config.js|https://webpack.js.org/concepts/configuration/>.

=head2 dependencies

  $hash_ref = $webpack->dependencies;

A hash where the keys can match the items in L</include> and the values are
lists of packages to install. Keys that does I</not> match items in L</include>
will be ignored. This attribute will be used by L</init>.

These dependencies are predefined:

  core   | webpack webpack-cli
  css    | css-loader mini-css-extract-plugin css-minimizer-webpack-plugin
  eslint | eslint-webpack-plugin
  js     | @babel/core @babel/preset-env @babel/plugin-transform-runtime babel-loader terser-webpack-plugin
  sass   | css-loader mini-css-extract-plugin css-minimizer-webpack-plugin sass sass-loader
  vue    | vue vue-loader vue-template-compiler

=head2 include

  $array_ref = $webpack->include;
  $webpack = $webpack->include([qw(js css)]);

L</include> can be used to install dependencies and load other webpack config
files. The config files included must exist in the "webpack.config.d" sub
directory inside L</assets_dir>. Here is an example of which files that will be
included if they exists:

  # Including "js" and "css" will look for the files below
  $webpack->include[qw(js css)]);

  # - assets/webpack.config.d/package-babel-loader.js
  # - assets/webpack.config.d/package-terser-webpack-plugin.js
  # - assets/webpack.config.d/package-css-loader.js
  # - assets/webpack.config.d/package-css-minimizer-webpack-plugin.js
  # - assets/webpack.config.d/js.js
  # - assets/webpack.config.d/css.js

The L</include> feature is currently EXPERIMENTAL.

=head2 mode

  $str = $webpack->mode;
  $webpack = $webpack->mode('development');

Should be either "development" or "production". Will be used as "NODE_ENV"
environment variable when calling L</build> or L</watch>.

=head2 npm

  $npm = $webpack->npm;

A L<Mojo::Alien::npm> object used by L</init>.

=head2 out_dir

  $path = $webpack->out_dir;
  $webpack = $webpack->out_dir(path('dist')->to_abs);

Location to write output assets to.

=head1 METHODS

=head2 asset_map

  $hash_ref = $webpack->asset_map;

Parses the filenames in L</out_dir> and returns a hash ref with information
about the generated assets. Example return value:

  {
    'relatibe/output.development.js' => {            # Key is relative path to out_dir()
      ext   => 'css',                                # File extension
      mode  => 'development',                        # or "production"
      mtime => 1616976114,                           # File modification epoch timestamp
      name  => 'relative/output.js',                 # Name of asset, without checksum or mode
      path  => '/path/to/entry-name.development.js', # Absolute path to asset
    },
  }

Note that this method is currently EXPERIMENTAL.

=head2 build

  $webpack->build;

Will build the assets or croaks on errors. Automatically calls L</init>.

=head2 exec

  $webpack->exec;

This method will replace the current process, instead of starting webpack
inside a fork. This method is called by L</watch> inside a fork.

=head2 init

  $webpack = $webpack->init;

Will install "webpack" and "webpack-cli" and create a default L</config>. Does
nothing if this is already done.

This method is automatically called by L</build> and L</watch>.

=head2 pid

  $int = $webpack->pid;

Returns the PID of the webpack process started by L</start>.

=head2 stop

  $webpack->stop;

Will stop the process started by L</watch>. Does nothing if L</watch> has not
been called.

=head2 watch

  $webpack->watch;

Forks a new process that runs "webpack watch". This means that any changes will
generate new assets. This is much more efficient than calling L</build> over
and over again. Automatically calls L</init>.

=head1 SEE ALSO

L<Mojolicious::Plugin::Webpack> and L<Mojo::Alien::rollup>.

=cut

__DATA__
@@ include/css.js
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const OptimizeCSSAssetsPlugin = require('css-minimizer-webpack-plugin');

module.exports = function(config, {isDev}) {
  if (!isDev) config.optimization.minimizer.push(new OptimizeCSSAssetsPlugin({}));
  config.plugins.push(new MiniCssExtractPlugin({filename: isDev ? '[name].development.css' : '[name].[contenthash].css'}));
  config.module.rules.push({
    test: /\.css$/,
    use: [MiniCssExtractPlugin.loader, {loader: 'css-loader', options: {sourceMap: true, url: false}}],
  });
};
@@ include/eslint.js
const ESLintPlugin = require('eslint-webpack-plugin');

module.exports = function(config) {
  config.plugins.push(new ESLintPlugin({
    exclude: ['node_modules/**', '**/*.css', '**/*.sass'],
    fix: process.env.ESLINT_FIX ? true : false,
  }));
};
@@ include/js.js
const TerserPlugin = require('terser-webpack-plugin');

module.exports = function(config, {isDev}) {
  if (!isDev) config.optimization.minimizer.push(new TerserPlugin({parallel: true}));

  config.module.rules.push({
    test: /\.js$/,
    exclude: /node_modules/,
    use: {
      loader: 'babel-loader',
      options: {
        plugins: ['@babel/plugin-transform-runtime'],
        presets: ['@babel/preset-env'],
      },
    },
  });
};
@@ include/sass.js
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const OptimizeCSSAssetsPlugin = require('css-minimizer-webpack-plugin');

module.exports = function(config, {isDev}) {
  if (!isDev) config.optimization.minimizer.push(new OptimizeCSSAssetsPlugin({}));
  config.plugins.push(new MiniCssExtractPlugin({filename: isDev ? '[name].development.css' : '[name].[contenthash].css'}));
  config.module.rules.push({
    test: /\.s(a|c)ss$/,
    use: [
      MiniCssExtractPlugin.loader,
      {loader: 'css-loader', options: {sourceMap: true, url: false}},
      {loader: 'sass-loader', options: {sourceMap: true}},
    ],
  });
};
@@ include/vue.js
const {VueLoaderPlugin} = require('vue-loader');

module.exports = function(config) {
  config.plugins.push(new VueLoaderPlugin());
  config.module.rules.push({test: /\.vue$/, use: 'vue-loader'});
};
@@ webpack.config.js
const fs = require('fs');
const pkg = require('./package.json');
const path = require('path');

const assetsDir = process.env.WEBPACK_ASSETS_DIR || path.resolve(__dirname, 'assets');
const isDev = process.env.NODE_ENV !== 'production';

const config = {
  entry: {},
  mode: isDev ? 'development' : 'production',
  module: {rules: []},
  optimization: {minimizer: []},
  output: {},
  plugins: [],
};

config.output.filename = isDev ? '[name].development.js' : '[name].[chunkhash].js';
config.output.path = process.env.WEBPACK_OUT_DIR || path.resolve(__dirname, 'dist');
config.output.publicPath = '';

const entry = path.resolve(assetsDir, 'index.js');
if (fs.existsSync(entry)) config.entry[pkg.name.replace(/\W+/g, '-')] = entry;

const includeFile = path.resolve(assetsDir, 'webpack.config.d', 'include.js');
if (fs.existsSync(includeFile)) require(includeFile)(config, {isDev});

// Legacy
const custom = path.resolve(assetsDir, 'webpack.custom.js');
if (fs.existsSync(custom)) require(custom)(config);

module.exports = config;
